tools/linkage_disequilibrium/ld_tool

```diff
@@ -0,0 +1,401 @@
+  foreach my $region (@regions) {
+    my ($chromosome, $start, $end) = split /\s+/, $region;
+    my $slice = $slice_adaptor->fetch_by_region('chromosome', $chromosome, $start, $end);
+    if ($slice) {
+      foreach my $population (@populations) {
+        my $ld_feature_container;
+        try {
+          $ld_feature_container = $ld_feature_container_adaptor->fetch_by_Slice($slice, $population);
+        } catch {
+          die "Error occurred during LD calculation: $_";
+        };
+        ld_feature_container_2_file($config, $ld_feature_container, $population);
+      }
+    } else {
+      &warnings_2_file($config, "Could not fetch region $region.");
+    }
+  }
+}
+
+sub center {
+  my $config = shift;
+  my @variants = @{parse_input($config)};
+  my @populations = @{$config->{_populations}};
+  my $ld_feature_container_adaptor = $config->{ld_feature_container_adaptor};
+  foreach my $variant (@variants) {
+    my $vf = get_variation_feature($config, $variant);
+    if ($vf) {
+      foreach my $population (@populations) {
+        my $ld_feature_container;
+        try {
+          $ld_feature_container = $ld_feature_container_adaptor->fetch_by_VariationFeature($vf, $population);
+        } catch {
+          die "Error occurred during LD caclculation.";
+        };
+        ld_feature_container_2_file($config, $ld_feature_container, $population);
+      }
+    } else {
+      &warnings_2_file($config, "Could not run LD calculation for $variant. Possible reason could be that variant doesn't have a location?");
+    }
+  }
+}
+
+sub parse_input {
+  my $config = shift;
+  my @input = ();
+  my $file = $config->{input_file};
+  if ($file) {
+    my $fh = FileHandle->new($file, 'r');
+    while (<$fh>) {
+      chomp;
+      s/^\s+|(\s+|\R)$//g;
+      push @input, $_ if ($_ ne '');
+    }
+    $fh->close;
+  } else {
+    if ($config->{calculation} eq 'region') {
+      @input = map {s/:|-/ /g} @regions;
+    } else {
+      @input = @variants;
+    }
+  }
+  my @uniq_input = uniq @input;
+  return \@uniq_input;
+}
+
+sub get_variation_feature {
+  my $config = shift;
+  my $variant = shift; 
+  my $variation_adaptor = $config->{variation_adaptor};
+
+  my $variation = $variation_adaptor->fetch_by_name($variant); 
+  if (!$variation) {
+    &warnings_2_file($config, "Could not fetch variant for $variant.");
+    return undef;
+  }   
+  my @variation_features = @{$variation->get_all_VariationFeatures};
+  my $vf = $variation_features[0];
+  if (scalar @variation_features > 1) {
+    my $chrom = $vf->seq_region_name; 
+    my $start = $vf->seq_region_start; 
+    my $end = $vf->seq_region_end;
+    &warnings_2_file($config, "Variation $variant has more than 1 mapping to the genome. Selected  $chrom:$start-$end as representative mapping.");
+  }
+  return $vf;
+}
+
+sub warnings_2_file {
+  my $config = shift;
+  my $warnings = shift;
+  my $warnings_file = $config->{warnings_file};
+  open(my $fh, '>>', $warnings_file);
```

Comment:
You sometimes see people trying to run scripts in directories they don't have write access to, so a-
'||die "Failed to open $warnings_file :$!\n'
type check/warning can be helpful.